src/App.tsx

```diff
@@ -1,61 +1,143 @@
 import './App.scss';
+import Button from 'react-bootstrap/Button';
+import Form from 'react-bootstrap/Form';
 
-// import usersFromServer from './api/users';
-// import todosFromServer from './api/todos';
+import { useState } from 'react';
+import { User } from './types/User';
+import { Todo } from './types/Todo';
+import { TodoList } from './components/TodoList';
+
+import usersFromServer from './api/users';
+import todosFromServer from './api/todos';
+
+function getUser(userId: number): User | null {
+  const foundUser = usersFromServer.find((user) => user.id === userId);
+
+  return foundUser || null;
+}
+
+function getMaximalId(todoes: Todo[]): number {
+  return Math.max(...todoes.map((todo) => todo.id)) + 1;
+}
+
+export const todoesList: Todo[] = [...todosFromServer].map((todo) => ({
+  ...todo,
+  user: getUser(todo.userId),
+}));
 
 export const App = () => {
+  const [title, setTitle] = useState('');
+  const [userId, setUserId] = useState<number>(0);
+  const [todoes, setNewTodo] = useState(todoesList);
+  const [isHaveTitle, setIsHaveTitle] = useState(false);
+  const [isHaveUserId, setIsHaveUserId] = useState(false);
+
+  const resetState = () => {
+    setTitle('');
+    setUserId(0);
+    setIsHaveTitle(false);
+    setIsHaveUserId(false);
+  };
+
+  const addNewTodo = (event: React.ChangeEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    const newTodo: Todo = {
+      id: getMaximalId(todoes),
+      title,
+      completed: false,
+      userId,
+      user: getUser(userId),
+    };
+
+    if (!title.length) {
+      setIsHaveTitle(true);
+    }
+
+    if (!userId) {
+      setIsHaveUserId(true);
+    }
+
+    if (title.length && userId) {
+      setNewTodo([...todoes, newTodo]);
```

Comment:
In ReactJS, it’s recommended to use functional `setState` instead of the regular `setState`. The reason for this is that the functional form of `setState` can help avoid issues with stale state, and ensure that the state is updated correctly.

Here are some rules to consider when using functional `setState` in React:

1. **Always use functional `setState` when updating state based on previous state:** If you need to update the state based on its previous value, it’s recommended to use functional `setState` to ensure that the state is updated correctly. Here’s an example:

```jsx
// using functional setState
this.setState(prevState => ({ count: prevState.count + 1 }));
```

2. **Use regular `setState` for simple updates:** If you’re updating the state with a simple value, such as a boolean or a string, you can use regular `setState`. Here’s an example:

```jsx
// using regular setState
this.setState({ isModalOpen: true });
```

3. **Use functional `setState` for complex state updates:** If you’re updating the state with a complex value, such as an array or an object, it’s recommended to use functional `setState` to ensure that the state is updated correctly. Here’s an example:

```jsx
// using functional setState with object spread syntax
this.setState(prevState => ({ ...prevState, items: newItems }));
```

4. **Be careful with nested state updates:** If you have nested state updates, it’s important to use functional `setState` to ensure that the state is updated correctly. Here’s an example:

```jsx
// incorrect use of setState with nested state updates
this.setState({ items: [] });
this.setState(prevState => ({ items: [...prevState.items, newItem] }));

// correct use of functional setState with nested state updates
this.setState(prevState => ({ items: [...prevState.items, newItem] }));
```

In summary, using functional `setState` in ReactJS can help avoid issues with stale state and ensure that the state is updated correctly. Always use functional `setState` when updating state based on its previous value, and be careful with nested state updates.